script/state_manager.js

```diff
@@ -0,0 +1,440 @@
+		}
+		if(!noEvent){
+			Engine.saveGame();
+			$SM.fireUpdate(stateName);
+		}
+	},
+
+	removeBranch: function(stateName, noEvent) {
+		for(var i in $SM.get(stateName)){
+			if(typeof $SM.get(stateName)[i] == 'object'){
+				$SM.removeBranch(stateName +'["'+ i +'"]');
+			}
+		}
+		if($.isEmptyObject($SM.get(stateName))){
+			$SM.remove(stateName);
+		}
+		if(!noEvent){
+			Engine.saveGame();
+			$SM.fireUpdate(stateName);
+		}
+	},
+
+	//creates full reference from input
+	//hopefully this won't ever need to be more complicated
+	buildPath: function(input){
+		var dot = (input.charAt(0) == '[')? '' : '.'; //if it starts with [foo] no dot to join
+		return 'State' + dot + input;
+	},
+
+	fireUpdate: function(stateName, save){
+		var category = $SM.getCategory(stateName);
+		if(stateName === undefined) stateName = category = 'all'; //best if this doesn't happen as it will trigger more stuff
+		$.Dispatch('stateUpdate').publish({'category': category, 'stateName':stateName});
+		if(save) Engine.saveGame();
+	},
+
+	getCategory: function(stateName){
+		var firstOB = stateName.indexOf('[');
+		var firstDot = stateName.indexOf('.');
+		var cutoff = null;
+		if(firstOB == -1 || firstDot == -1){
+			cutoff = firstOB > firstDot ? firstOB : firstDot;
+		} else {
+			cutoff = firstOB < firstDot ? firstOB : firstDot;
+		}
+		if (cutoff == -1){
+			return stateName;
+		} else {
+			return stateName.substr(0,cutoff);
+		}
+	},
+
+	//Use this function to make old save games compatible with new version
+	updateOldState: function(){
+		var version = $SM.get('version');
+		if(typeof version != 'number') version = 1.0;
+		if(version == 1.0) {
+			// v1.1 introduced the Lodge, so get rid of lodgeless hunters
+			$SM.remove('outside.workers.hunter', true);
+			$SM.remove('income.hunter', true);
+			Engine.log('upgraded save to v1.1');
+			version = 1.1;
+		}
+		if(version == 1.1) {
+			//v1.2 added the Swamp to the map, so add it to already generated maps
+			if($SM.get('world')) {
+				World.placeLandmark(15, World.RADIUS * 1.5, World.TILE.SWAMP, $SM.get('world.map'));
+			}
+			Engine.log('upgraded save to v1.2');
+			version = 1.2;
+		}
+		if(version == 1.2) {
+			//StateManager added, so move data to new locations
+			$SM.remove('room.fire');
+			$SM.remove('room.temperature');
+			$SM.remove('room.buttons');
+			if($SM.get('room')){
+				$SM.set('features.location.room', true);
+				$SM.set('game.builder.level', $SM.get('room.builder'));
+				$SM.remove('room');
+			}
+			if($SM.get('outside')){
+				$SM.set('features.location.outside', true);
+				$SM.set('game.population', $SM.get('outside.population'));
+				$SM.set('game.buildings', $SM.get('outside.buildings'));
+				$SM.set('game.workers', $SM.get('outside.workers'));
+				$SM.set('game.outside.seenForest', $SM.get('outside.seenForest'));
+				$SM.remove('outside');
+			}
+			if($SM.get('world')){
+				$SM.set('features.location.world', true);
+				$SM.set('game.world.map', $SM.get('world.map'));
```

Comment:
This if statement calls the set function multiple times. Rather than writing each set function separately, a for loop could be written to call each statement based on the number in the loop. Each loop number could correspond with the variables that need to be passed in.